pilot/pkg/xds/debug.go

```diff
@@ -672,91 +673,157 @@ func (s *DiscoveryServer) ConfigDump(w http.ResponseWriter, req *http.Request) {
+		})
+	}
+	routesAny, err := util.MessageToAnyWithError(&adminapi.RoutesConfigDump{
+		DynamicRouteConfigs: dynamicRouteConfig,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	secrets, err := generate(v3.SecretType)
+	if err != nil {
+		return nil, err
+	}
+	dynamicSecretsConfig := make([]*adminapi.SecretsConfigDump_DynamicSecret, 0)
+	for _, cs := range secrets {
+		// Secrets must be redacted
+		secret := &tls.Secret{}
+		if err := cs.Resource.UnmarshalTo(secret); err != nil {
+			istiolog.Warnf("failed to unmarshal secret: %v", err)
+			continue
 		}
-		routeConfigAny, err = util.MessageToAnyWithError(&adminapi.RoutesConfigDump{DynamicRouteConfigs: dynamicRouteConfig})
-		if err != nil {
-			return nil, err
+		if secret.GetTlsCertificate() != nil {
+			secret.GetTlsCertificate().PrivateKey = &core.DataSource{
+				Specifier: &core.DataSource_InlineBytes{
+					InlineBytes: []byte("[redacted]"),
+				},
+			}
 		}
+		dynamicSecretsConfig = append(dynamicSecretsConfig, &adminapi.SecretsConfigDump_DynamicSecret{
+			VersionInfo: version,
+			Secret:      util.MessageToAny(secret),
+		})
+	}
+	secretsAny, err := util.MessageToAnyWithError(&adminapi.SecretsConfigDump{
+		DynamicActiveSecrets: dynamicSecretsConfig,
+	})
+	if err != nil {
+		return nil, err
 	}
 
-	secretsDump := &adminapi.SecretsConfigDump{}
-	if s.Generators[v3.SecretType] != nil {
-		secrets, _, _ := s.Generators[v3.SecretType].Generate(conn.proxy, conn.Watched(v3.SecretType), nil)
-		if len(secrets) > 0 {
-			for _, secretAny := range secrets {
-				secret := &tls.Secret{}
-				if err := secretAny.GetResource().UnmarshalTo(secret); err != nil {
-					istiolog.Warnf("failed to unmarshal secret: %v", err)
-				}
-				if secret.GetTlsCertificate() != nil {
-					secret.GetTlsCertificate().PrivateKey = &core.DataSource{
-						Specifier: &core.DataSource_InlineBytes{
-							InlineBytes: []byte("[redacted]"),
-						},
-					}
-				}
-				secretsDump.DynamicActiveSecrets = append(secretsDump.DynamicActiveSecrets, &adminapi.SecretsConfigDump_DynamicSecret{
-					Name:   secret.Name,
-					Secret: util.MessageToAny(secret),
-				})
-			}
+	var endpointsAny *any.Any
+	// EDS is disabled by default for compatibility with Envoy config_dump interface
+	if includeEds {
+		endpoints, err := generate(v3.EndpointType)
+		if err != nil {
+			return nil, err
+		}
+		endpointConfig := make([]*adminapi.EndpointsConfigDump_DynamicEndpointConfig, 0)
+		for _, cs := range endpoints {
+			endpointConfig = append(endpointConfig, &adminapi.EndpointsConfigDump_DynamicEndpointConfig{
+				VersionInfo:    version,
+				EndpointConfig: cs.Resource,
+			})
+		}
+		endpointsAny, err = util.MessageToAnyWithError(&adminapi.EndpointsConfigDump{
+			DynamicEndpointConfigs: endpointConfig,
+		})
+		if err != nil {
+			return nil, err
 		}
 	}
 
 	bootstrapAny := util.MessageToAny(&adminapi.BootstrapConfigDump{})
 	scopedRoutesAny := util.MessageToAny(&adminapi.ScopedRoutesConfigDump{})
```

Comment:
why have these empty structs

Reply:
The idea is this endpoint behaves exactly like Envoy's config dump, which includes these. Since we never provide these types we just always generate empty

BTW this part of the code is unchanged.